db/migrations/25-03-07-compute-contribution-stats.sql

```diff
@@ -0,0 +1,133 @@
+    language_id uuid not null references language (id),
+    user_id uuid not null references users (id),
+    approved_count int not null,
+    revoked_count int not null,
+    edited_approved_count int not null,
+    edited_unapproved_count int not null,
+    unique (week, language_id, user_id)
+);
+
+create or replace function generate_weekly_contribution_stats(d timestamp)
+returns void
+language sql
+as $$
+    insert into weekly_contribution_statistics (
+        week,
+        language_id,
+        user_id,
+        approved_count,
+        revoked_count,
+        edited_approved_count,
+        edited_unapproved_count
+    )
+    select
+        date_bin('7 days', date_trunc('day', $1), timestamp '2024-12-15'),
+        changed_phrase.language_id,
+        changed_phrase.updated_by,
+        coalesce(count(*) filter (
+            where changed_phrase.state = 'APPROVED'
+                and coalesce(prev_phrase.state, 'UNAPPROVED') = 'UNAPPROVED'
+        ), 0) as words_approved,
+        coalesce(count(*) filter (
+            where changed_phrase.state = 'UNAPPROVED'
+                and coalesce(prev_phrase.state, 'UNAPPROVED') = 'APPROVED'
+        ), 0) as words_revoked,
+        coalesce(count(*) filter (
+            where changed_phrase.gloss <> coalesce(prev_phrase.gloss, '')
+                and changed_phrase.state = 'APPROVED'
+                and coalesce(prev_phrase.state, 'UNAPPROVED') = 'APPROVED'
+        ), 0) as words_approved_edited,
+        coalesce(count(*) filter (
+            where changed_phrase.gloss <> coalesce(prev_phrase.gloss, '')
+                and changed_phrase.state = 'UNAPPROVED'
+                and coalesce(prev_phrase.state, 'UNAPPROVED') = 'UNAPPROVED'
+        ), 0) as words_unapproved_edited
+    from (
+        select
+            DISTINCT ON (ph.language_id, phw.word_id)
+            ph.language_id, phw.word_id,
+            log.*
+        from (
+            (
+                select phrase_id, updated_by, updated_at, gloss, state
+                from gloss
+            ) union all (
+                select phrase_id, updated_by, updated_at, gloss, state
+                from gloss_history
+            ) union all (
+                select
+                    id as phrase_id,
+                    deleted_by as updated_by,
+                    deleted_at as updated_at,
+                    '' as gloss,
+                    'UNAPPROVED' as state
+                from phrase
+                where deleted_by is not null and deleted_at is not null
+            )
+        ) log
+        join phrase_word phw on phw.phrase_id = log.phrase_id
+        join phrase ph on ph.id = log.phrase_id
+        where log.updated_at > date_bin('7 days', date_trunc('day', $1), timestamp '2024-12-15')
+            and log.updated_at < date_bin('7 days', date_trunc('day', $1 + interval '7 days'), timestamp '2024-12-15')
+        order by ph.language_id, phw.word_id, log.updated_at desc
```

Comment:
This chunk of the query grabs the last change made for each word, in each language, during the week in question. This also includes when phrases are deleted during import or word linking